File: Hosts/VS/TddStudioPackage/Services.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.Shell;
using R4nd0mApps.TddStud10.Hosts.VS.Diagnostics;

namespace R4nd0mApps.TddStud10.Hosts.VS
{
    // TODO: Move to fs
    public static class Services
    {
        public static TI GetService<TS, TI>(this IServiceProvider serviceProvider)
        {
            var s = serviceProvider.GetService(typeof(TS));
            if (s == null)
            {
                Logger.I.LogError("Unable to query for service of type {0}.", typeof(TS).FullName);
            }

            TI svc = default(TI);
            try
            {
                svc = (TI)s;
            }
            catch
            {
                Logger.I.LogError("Cannot cast service '{0}' to '{1}'.", typeof(TS).FullName, typeof(TI).FullName);
            }

            return svc;
        }

        public static T GetService<T>()
        {
            return GetService<T, T>();
        }

        public static TI GetService<TS, TI>()
        {
            return ServiceProvider.GlobalProvider.GetService<TS, TI>();
        }

    }
}

File: Hosts/VS/TddStudioPackage.Extensions/Services.fs
namespace R4nd0mApps.TddStud10.Hosts.VS.TddStudioPackage.Extensions

open System
open System.Runtime.CompilerServices
open R4nd0mApps.TddStud10.Hosts.VS.Diagnostics
open Microsoft.VisualStudio.Shell

// NOTE: We should ideally return Option<'T> from these functions. Do it when we get a chance next.
[<Extension>]
type public Services = 
    
    [<Extension>]
    static member public GetService<'TS, 'TI when 'TI : null>(sp : IServiceProvider) : 'TI = 
        let s = sp.GetService(typeof<'TS>)
        match s with
        | null -> 
            Logger.logErrorf "Unable to query for service of type %s." typeof<'TS>.FullName
            null
        | :? 'TI as i -> i
        | _ -> 
            Logger.logErrorf "Cannot cast service '%s' to '%s'." typeof<'TS>.FullName typeof<'TI>.FullName
            null

    [<Extension>]
    static member public GetService<'T when 'T : null>(sp : IServiceProvider) : 'T = 
        Services.GetService<'T, 'T>(sp)

    static member public GetService<'T when 'T : null>() : 'T = 
        Services.GetService<'T, 'T>(ServiceProvider.GlobalProvider)

    static member public GetService<'TS, 'TI when 'TI : null>() : 'TI = 
        Services.GetService<'TS, 'TI>(ServiceProvider.GlobalProvider)
